### PR TITLE
Fix frontend_server problem with _filename not being set in `compile`.

### PR DESCRIPTION
**Pull request by aam**
And now that mockito is on chromium source mirror, switch to mockito for all existing tests, add new test for _filename issue.

### CODE REVIEW

File: frontend_server/lib/server.dart [hunk 1]
```diff
@@ -59,8 +59,10 @@ enum _State { READY_FOR_INSTRUCTION, RECOMPILE_LIST }
 /// Actions that every compiler should implement.
 abstract class CompilerInterface {
   /// Compile given Dart program identified by `filename` with given list of
-  /// `options`.
-  Future<Null> compile(String filename, ArgResults options);
+  /// `options`. If `generator` instance is provided(primarily for mocking in
```
Comment (Hixie) [inline frontend_server/lib/server.dart]: nit: space before `(`

Comment (Hixie) [inline frontend_server/lib/server.dart]: also, you're missing a `)`. Not sure exactly what you're trying to say here.

File: frontend_server/lib/server.dart [hunk 2]
```diff
@@ -59,8 +59,10 @@ enum _State { READY_FOR_INSTRUCTION, RECOMPILE_LIST }
...
+  /// tests, it is used for compilation. Otherwise new instance will be created.
+  Future<Null> compile(String filename, ArgResults options,
+    {IncrementalKernelGenerator generator});
```
Comment (Hixie) [inline frontend_server/lib/server.dart]:
```dart
  Future<Null> compile(String filename, ArgResults options, {
    IncrementalKernelGenerator generator,
  });
```

File: frontend_server/lib/server.dart [hunk 3]
```diff
@@ -78,17 +80,31 @@ abstract class CompilerInterface {
...
+class KernelSerializer {
+  /// Writes given kernel `program` to `path`.
+  Future<dynamic> writeProgramToBinary(Program program, String path) {
+    return kernel.writeProgramToBinary(program, path);
+  }
+}
+
 class _FrontendCompiler implements CompilerInterface {
-  _FrontendCompiler(this._outputStream) {
+  _FrontendCompiler(this._outputStream, {this.kernelSerializer}) {
```
Comment (Hixie) [inline frontend_server/lib/server.dart]: nit: spaces around argument names

File: frontend_server/lib/server.dart [hunk 4]
```diff
@@ -78,17 +80,31 @@ abstract class CompilerInterface {
...
     _outputStream ??= stdout;
+    kernelSerializer ??= new KernelSerializer();
   }
 
+  KernelSerializer kernelSerializer;
+
   IncrementalKernelGenerator _generator;
   String _filename;
   StringSink _outputStream;
```
Comment (Hixie) [inline frontend_server/lib/server.dart]: nit: _outputStream should be above kernelSerializer, to match constructor order

File: frontend_server/lib/server.dart [hunk 5]
```diff
@@ -78,17 +80,31 @@ abstract class CompilerInterface {
...
 
   @override
-  Future<Null> compile(String filename, ArgResults options) async {
+  Future<Null> compile(String filename, ArgResults options,
+    {IncrementalKernelGenerator generator}) async {
```
Comment (Hixie) [inline frontend_server/lib/server.dart]: nit: similar indentation to the suggestion above

File: frontend_server/lib/server.dart [hunk 6]
```diff
@@ -101,9 +117,10 @@ class _FrontendCompiler implements CompilerInterface {
       };
     Program program;
     if (options['incremental']) {
-      _generator = await IncrementalKernelGenerator.newInstance(
-          compilerOptions, Uri.base.resolve(filename)
-      );
+      _generator = generator != null
+        ? generator
+        : await IncrementalKernelGenerator.newInstance(
+            compilerOptions, Uri.base.resolve(_filename));
```
Comment (Hixie) [inline frontend_server/lib/server.dart]: nit: if you have a newline after the (, you need one before the )

File: frontend_server/lib/server.dart [hunk 7]
```diff
@@ -163,16 +179,18 @@ class _FrontendCompiler implements CompilerInterface {
 /// `compiler` is an optional parameter so it can be replaced with mocked
 /// version for testing.
 Future<int> starter(List<String> args, {CompilerInterface compiler,
-    Stream<List<int>> input, StringSink output}) async {
+    Stream<List<int>> input, StringSink output,
+    IncrementalKernelGenerator generator,
+    KernelSerializer kernelSerializer}) async {
```
Comment (Hixie) [inline frontend_server/lib/server.dart]: nit: style here too, similar as above (`{` at the end of the first line, then line break, if it contains any line breaks, and line break before the `}` if there's one after the `{`)

**Comment by Hixie** [comment]
Man, I really find mockito-based tests hard to read. But whatever it takes to get more tests. :-)

LGTM modulo style nits.

**Comment by aam** [comment]
> Man, I really find mockito-based tests hard to read. But whatever it takes to get more tests. :-)

Yeah, good to hear confirmation of my own sentiment towards mockito(this is my first attempt at using it in dart). I think it lose its appeal with async-based interfaces, or when you want more complex checks implemented. Might consider reverting back to original hand-made mocking.

File: frontend_server/lib/server.dart (hunk 7 above)
Comment (aam) [inline frontend_server/lib/server.dart]: Neat, done.

File: frontend_server/lib/server.dart (hunk 1 above)
Comment (aam) [inline frontend_server/lib/server.dart]: Reworded!

**Comment by aam** [comment]
@sigmundch , realized too late that you might have some input regarding how I use mockito package since you mentioned it on one of the recent PRs. As it stands it feels a little clunky, which might be an "operator's fault".

**Comment by sigmundch** [comment]
no worries, not much input on my end about mockito - I only know about it by proxy , but haven't used it myself :)

**Pull request merged**